Cython/Compiler/ExprNodes.py

```diff
@@ -3119,19 +3119,32 @@ def generate_next_sequence_item(self, test_name, result_name, code):
             inc_dec = '--'
         else:
             inc_dec = '++'
-        code.putln("#if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS")
-        code.putln(
-            "%s = Py%s_GET_ITEM(%s, %s); __Pyx_INCREF(%s); %s%s; %s" % (
-                result_name,
-                test_name,
-                self.py_result(),
-                self.counter_cname,
-                result_name,
-                self.counter_cname,
-                inc_dec,
-                # use the error label to avoid C compiler warnings if we only use it below
-                code.error_goto_if_neg('0', self.pos)
-                ))
+        if test_name == 'List':
+            code.putln("#if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS")
+            code.putln(
+                "%s = __Pyx_PyList_GetItemRef(%s, %s); __Pyx_GOTREF(%s); %s%s; %s" % (
+                    result_name,
+                    self.py_result(),
+                    self.counter_cname,
+                    result_name,
+                    self.counter_cname,
+                    inc_dec,
+                    # use the error label to avoid C compiler warnings if we only use it below
+                    code.error_goto_if_neg('0', self.pos)
+                    ))
+        else:  # Tuple
+            code.putln("#if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS")
+            code.putln(
+                "%s = PyTuple_GET_ITEM(%s, %s); __Pyx_INCREF(%s); %s%s; %s" % (
+                    result_name,
+                    self.py_result(),
+                    self.counter_cname,
+                    result_name,
+                    self.counter_cname,
+                    inc_dec,
+                    # use the error label to avoid C compiler warnings if we only use it below
+                    code.error_goto_if_neg('0', self.pos)
+                    ))
```

Comment:
It would be easier to spot the differences between both cases if you limited the conditional code to just those, without repeating the entire code generation. I'd decide between the two macro names and then use them to do the code generation outside of the if-else blocks.